File: src/engine/adapters/redis/console.ts
import {
  createKeyValuePageBuilder,
  type KeyValuePage,
  type Page,
  type PagePosition,
} from '@shared/protocol/page';
import type { OpCtx } from '../adapter';
import { AdapterError } from '../errors';
import type { DbConnectionSet } from './client';
import { mapError } from './errors';

// §8.14: "for non-SQL engines the console takes that engine's native command form" — real Redis
// CLI syntax is flat whitespace-separated tokens with optional single/double quoting (backslash
// escapes inside quotes), not a JSON DSL like Mongo's shell (P9's D11).
function tokenize(line: string): string[] {
  const tokens: string[] = [];
  const n = line.length;
  let i = 0;
  while (i < n) {
    while (i < n && /\s/.test(line[i])) i++;
    if (i >= n) break;
    const quote = line[i];
    if (quote === '"' || quote === "'") {
      i++;
      let out = '';
      while (i < n && line[i] !== quote) {
        if (line[i] === '\\' && i + 1 < n) {
          out += line[i + 1];
          i += 2;
        } else {
          out += line[i];
          i++;
        }
      }
      if (i >= n) throw new AdapterError('E_QUERY', 'unterminated quoted string');
      i++; // closing quote
      tokens.push(out);
    } else {
      let out = '';
      while (i < n && !/\s/.test(line[i])) {
        out += line[i];
        i++;
      }
      tokens.push(out);
    }
  }
  return tokens;
}

function formatReplyItem(value: unknown): string {
  if (value === null || value === undefined) return '(nil)';
  if (typeof value === 'string') return value;
  if (typeof value === 'number' || typeof value === 'boolean') return String(value);
  if (Buffer.isBuffer(value)) return value.toString('utf8');
  return JSON.stringify(value);
}

// Any RESP reply is formatted generically (P9's D11) — no per-command result shape, unlike
// Mongo's console, which needs one because each shell method returns a different document shape.
function resultToPage(command: string, reply: unknown): KeyValuePage {
  const builder = createKeyValuePageBuilder({
    redisType: 'string',
    ttlMs: null,
    memoryBytes: null,
  });
  let pageSize: number;
  if (Array.isArray(reply)) {
    reply.forEach((item, i) => {
      builder.push(String(i), formatReplyItem(item));
    });
    pageSize = reply.length;
  } else {
    builder.push(command.toUpperCase(), formatReplyItem(reply));
    pageSize = 1;
  }
  const position: PagePosition = {
    offset: 0,
    pageSize,
    hasMore: false,
    nextToken: null,
    prevToken: null,
    strategy: 'offset',
  };
  return builder.finish(position);
}

// One op-log row for the whole batch (P5.5 D9's precedent, mirrored from mysql-family/console.ts and
// mongo/console.ts).
export async function execute(
  set: DbConnectionSet,
  dbIndex: number,
  ctx: OpCtx,
  statements: string[],
): Promise<Page[]> {
  const lines = statements.map((s) => s.trim()).filter((s) => s !== '');
  if (lines.length === 0) throw new AdapterError('E_QUERY', 'no statements to execute');
  ctx.setCommand(lines.join('\n'));

  const conn = await set.get(dbIndex);
  const pages: Page[] = [];
  for (const line of lines) {
    if (ctx.signal.aborted) throw new AdapterError('E_CANCELLED', 'operation was cancelled');
    const [command, ...args] = tokenize(line);
    if (!command) continue;
    try {
      const reply = await conn.call(command, ...args);
      pages.push(resultToPage(command, reply));
    } catch (err) {
      throw mapError(err);
    }
  }
  if (pages.length === 0) throw new AdapterError('E_QUERY', 'no statements to execute');
  return pages;
}

File: shell/internal/adapters/redis/console.go
package redis

import (
	"context"
	"encoding/json"
	"strconv"
	"strings"
	"unicode"

	"github.com/kirathecat/kira-studio/shell/internal/adapters"
	"github.com/kirathecat/kira-studio/shell/internal/page"
)

// §8.14: "for non-SQL engines the console takes that engine's native command form" — real Redis
// CLI syntax is flat whitespace-separated tokens with optional single/double quoting (backslash
// escapes inside quotes), not a JSON DSL like Mongo's shell (P9's D11, C12).
func tokenize(line string) ([]string, error) {
	runes := []rune(line)
	n := len(runes)
	var tokens []string
	i := 0
	for i < n {
		for i < n && unicode.IsSpace(runes[i]) {
			i++
		}
		if i >= n {
			break
		}
		quote := runes[i]
		if quote == '"' || quote == '\'' {
			i++
			var out strings.Builder
			for i < n && runes[i] != quote {
				if runes[i] == '\\' && i+1 < n {
					out.WriteRune(runes[i+1])
					i += 2
				} else {
					out.WriteRune(runes[i])
					i++
				}
			}
			if i >= n {
				return nil, adapters.New(adapters.CodeQuery, "unterminated quoted string", nil)
			}
			i++ // closing quote
			tokens = append(tokens, out.String())
		} else {
			var out strings.Builder
			for i < n && !unicode.IsSpace(runes[i]) {
				out.WriteRune(runes[i])
				i++
			}
			tokens = append(tokens, out.String())
		}
	}
	return tokens, nil
}

func formatReplyItem(value any) string {
	if value == nil {
		return "(nil)"
	}
	switch v := value.(type) {
	case string:
		return v
	case int64:
		return strconv.FormatInt(v, 10)
	case float64:
		return strconv.FormatFloat(v, 'g', -1, 64)
	case bool:
		return strconv.FormatBool(v)
	default:
		b, err := json.Marshal(v)
		if err != nil {
			return ""
		}
		return string(b)
	}
}

// resultToPage is console.ts's own: any RESP reply is formatted generically (P9's D11) — no
// per-command result shape, unlike Mongo's console. A scalar's field name is the upper-cased
// command; an array's are the indices.
func resultToPage(command string, reply any) page.KeyValuePage {
	builder := page.NewKeyValuePageBuilder("string", nil, nil, false)
	pageSize := 1
	if arr, ok := reply.([]any); ok {
		for i, item := range arr {
			builder.Push(strconv.Itoa(i), formatReplyItem(item))
		}
		pageSize = len(arr)
	} else {
		builder.Push(strings.ToUpper(command), formatReplyItem(reply))
	}
	return builder.Finish(page.UnpagedPosition(pageSize))
}

// execute is console.ts's execute — one op-log row for the whole batch (P5.5 D9's precedent).
func execute(ctx context.Context, set *dbConnectionSet, dbIndex int, op *adapters.OpCtx, statements []string) ([]page.Page, error) {
	var lines []string
	for _, s := range statements {
		if trimmed := strings.TrimSpace(s); trimmed != "" {
			lines = append(lines, trimmed)
		}
	}
	if len(lines) == 0 {
		return nil, adapters.New(adapters.CodeQuery, "no statements to execute", nil)
	}
	op.SetCommand(strings.Join(lines, "\n"))

	conn, err := set.get(ctx, dbIndex)
	if err != nil {
		return nil, err
	}

	var pages []page.Page
	for _, line := range lines {
		if err := adapters.CheckCancelled(ctx); err != nil {
			return nil, err
		}
		tokens, err := tokenize(line)
		if err != nil {
			return nil, err
		}
		if len(tokens) == 0 {
			continue
		}
		command, args := tokens[0], tokens[1:]
		argv := make([]any, 0, len(args)+1)
		argv = append(argv, command)
		for _, a := range args {
			argv = append(argv, a)
		}
		reply, err := conn.Do(ctx, argv...).Result()
		if err != nil {
			return nil, mapError(err)
		}
		pages = append(pages, resultToPage(command, reply))
	}
	if len(pages) == 0 {
		return nil, adapters.New(adapters.CodeQuery, "no statements to execute", nil)
	}
	return pages, nil
}
